File: 3399.smallest-substring-with-identical-characters-ii.java
#
# @lc app=leetcode id=3399 lang=java
#
# [3399] Smallest Substring With Identical Characters II
#

# @lc code=start
class Solution {
    public int minLength(String s, int numOps) {
        int n = s.length();
        int left = 1, right = n;
        while (left < right) {
            int mid = left + (right - left) / 2;
            if (canAchieve(s, numOps, mid)) {
                right = mid;
            } else {
                left = mid + 1;
            }
        }
        return left;
    }

    private boolean canAchieve(String s, int numOps, int maxLen) {
        long need1 = simulate(s, maxLen, false);
        if (need1 <= numOps) {
            return true;
        }
        long need2 = simulate(s, maxLen, true);
        return need2 <= numOps;
    }

    private long simulate(String s, int maxLen, boolean flipFirst) {
        int n = s.length();
        long need = flipFirst ? 1L : 0L;
        int currCh = flipFirst ? 1 - (s.charAt(0) - '0') : (s.charAt(0) - '0');
        int currLen = 1;
        for (int i = 1; i < n; ++i) {
            int origCh = s.charAt(i) - '0';
            int preferCh = origCh;
            if (preferCh == currCh && currLen == maxLen) {
                ++need;
                preferCh = 1 - origCh;
            }
            int chosenCh = preferCh;
            if (chosenCh == currCh) {
                ++currLen;
            } else {
                currCh = chosenCh;
                currLen = 1;
            }
        }
        return need;
    }
}
# @lc code=end

File: 3399.smallest-substring-with-identical-characters-ii.go
#
# @lc app=leetcode id=3399 lang=golang
#
# [3399] Smallest Substring With Identical Characters II
#
# @lc code=start
func minLength(s string, numOps int) int {
    n := len(s)
    
    // Helper function to check if maxLen is achievable with at most numOps flips
    canAchieve := func(maxLen int) bool {
        if maxLen == 1 {
            // Special case: need alternating pattern
            mismatch0 := 0 // Pattern "0101..."
            mismatch1 := 0 // Pattern "1010..."
            for i := 0; i < n; i++ {
                expected0 := byte('0' + i%2)
                expected1 := byte('1' - i%2)
                if s[i] != expected0 {
                    mismatch0++
                }
                if s[i] != expected1 {
                    mismatch1++
                }
            }
            return min(mismatch0, mismatch1) <= numOps
        }
        
        // For maxLen >= 2, count flips needed to break runs
        opsNeeded := 0
        i := 0
        for i < n {
            j := i
            // Find the end of current run
            for j < n && s[j] == s[i] {
                j++
            }
            runLen := j - i
            if runLen > maxLen {
                // Number of flips needed to ensure each segment is at most maxLen
                opsNeeded += runLen / (maxLen + 1)
            }
            i = j
        }
        return opsNeeded <= numOps
    }
    
    // Binary search on the answer
    left, right := 1, n
    result := n
    
    for left <= right {
        mid := (left + right) / 2
        if canAchieve(mid) {
            result = mid
            right = mid - 1
        } else {
            left = mid + 1
        }
    }
    
    return result
}

func min(a, b int) int {
    if a < b {
        return a
    }
    return b
}
# @lc code=end